<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Reads in spatial data in the form of cell coordinates, cell
  phenotypes (if available), and marker intensities and transforms to a
  SpatialExperiment object. The assay stores the intensity level of every
  marker (rows) for every cell (columns). Cell phenotype is stored under
  `colData()`. Cell x and y coordinates are stored under `spatialCoords()`
  field. The function can read in data in general format (manually
  constructed input), or data from other platforms including inForm, HALO,
  CODEX and cellprofiler. Alternatively, users can use the specific function
  for each format."><title>Format an image into a SpatialExperiment object — format_image_to_spe • SPIAT</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Format an image into a SpatialExperiment object — format_image_to_spe"><meta property="og:description" content="Reads in spatial data in the form of cell coordinates, cell
  phenotypes (if available), and marker intensities and transforms to a
  SpatialExperiment object. The assay stores the intensity level of every
  marker (rows) for every cell (columns). Cell phenotype is stored under
  `colData()`. Cell x and y coordinates are stored under `spatialCoords()`
  field. The function can read in data in general format (manually
  constructed input), or data from other platforms including inForm, HALO,
  CODEX and cellprofiler. Alternatively, users can use the specific function
  for each format."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SPIAT</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.4</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/introduction.html">Overview of the SPIAT package</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/trigosteam/SPIAT/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Format an image into a SpatialExperiment object</h1>
      <small class="dont-index">Source: <a href="https://github.com/trigosteam/SPIAT/blob/HEAD/R/format_image_to_spe.R" class="external-link"><code>R/format_image_to_spe.R</code></a></small>
      <div class="d-none name"><code>format_image_to_spe.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Reads in spatial data in the form of cell coordinates, cell
  phenotypes (if available), and marker intensities and transforms to a
  SpatialExperiment object. The assay stores the intensity level of every
  marker (rows) for every cell (columns). Cell phenotype is stored under
  `colData()`. Cell x and y coordinates are stored under `spatialCoords()`
  field. The function can read in data in general format (manually
  constructed input), or data from other platforms including inForm, HALO,
  CODEX and cellprofiler. Alternatively, users can use the specific function
  for each format.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">format_image_to_spe</span><span class="op">(</span></span>
<span>  format <span class="op">=</span> <span class="st">"general"</span>,</span>
<span>  intensity_matrix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  phenotypes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  coord_x <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  coord_y <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  markers <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  locations <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  intensity_columns_interest <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  dye_columns_interest <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  path_to_codex_cell_phenotypes <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>format</dt>
<dd><p>String specifying the format of the data source. Default is
"general" (RECOMMENDED), where the cell phenotypes, coordinates and marker
intensities are imported manually by the user. Other formats include
"inForm", "HALO", "cellprofiler" and "CODEX".</p></dd>


<dt>intensity_matrix</dt>
<dd><p>(Optional) For "general" format. A matrix of marker
intensities or gene expression where the column names are the Cell IDs, and
the rownames the marker.</p></dd>


<dt>phenotypes</dt>
<dd><p>(Optional) For "general" format. String Vector of cell
phenotypes in the same order in which they appear in `intensity_matrix`. If
no phenotypes available, then a vector of NAs can be used as input. Note
that the combination of markers (e.g. CD3,CD4) needs to be used instead of
the cell type name (e.g. helper T cells).</p></dd>


<dt>coord_x</dt>
<dd><p>(Optional) For "general" format. Numeric Vector with the X
coordinates of the cells. The cells must be in the same order as in the
`intensity_matrix`.</p></dd>


<dt>coord_y</dt>
<dd><p>(Optional) For "general" format. Numeric Vector with the Y
coordinates of the cells. The cells must be in the same order as in the
`intensity_matrix`.</p></dd>


<dt>path</dt>
<dd><p>(Optional) For formats other than "general". String of the path
location of the source file.</p></dd>


<dt>markers</dt>
<dd><p>For formats other than "general". String Vector containing the
markers used for staining. These must be in the same order as the marker
columns in the input file, and must match the marker names used in the
input file. One of the markers must be "DAPI".</p></dd>


<dt>locations</dt>
<dd><p>(Optional) For "inForm" and "HALO". String Vector containing
the locations of markers used for staining. Location can be either
"Nucleus", "Cytoplasm" or "Membrane". This is used to select the Intensity
column and can be used instead of `intensity_columns_interest`.</p></dd>


<dt>intensity_columns_interest</dt>
<dd><p>(Optional) For "inForm" and "HALO", use if
`locations` is not specified. For "cellprofiler", mandatory. Vector with
the names of the columns with the level of each marker. Column names must
match the order of the 'markers' parameter.</p></dd>


<dt>dye_columns_interest</dt>
<dd><p>(Optional) For "HALO". Use if locations is not
specified. Vector of names of the columns with the marker status (i.e.
those indicating 1 or 0 for whether the cell is positive or negative for
the marker). Column names must match the order of the 'markers' parameter.</p></dd>


<dt>path_to_codex_cell_phenotypes</dt>
<dd><p>(Optional) For "CODEX".String of the
path to the Cluster ID/Cell type file.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A SpatialExperiment object is returned</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Note for "cellprofiler" format, when specifying `markers`, please
  use "DAPI" to replace "DNA" due to implementation. The output data will
  include "DAPI" instead of "DNA".</p>
<p>The format of "Phenotype" column: For example, a cell positive for both
  "CD3" and "CD4" markers has the "CD3,CD4" **cell phenotype**. The phenotype
  has to be strictly formatted in such way where each positive marker has to
  be separated by a coma, with no space in between, and the order of the
  positive markers has to be the same as the order in the assay.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="format_inform_to_spe.html">format_inform_to_spe</a></code> <code><a href="format_halo_to_spe.html">format_halo_to_spe</a></code>
  <code><a href="format_codex_to_spe.html">format_codex_to_spe</a></code> <code><a href="format_cellprofiler_to_spe.html">format_cellprofiler_to_spe</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">#Construct a marker intensity matrix (rows are markers, columns are cells)</span></span></span>
<span class="r-in"><span><span class="va">intensity_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">14.557</span>, <span class="fl">0.169</span>, <span class="fl">1.655</span>, <span class="fl">0.054</span>, <span class="fl">17.588</span>, <span class="fl">0.229</span>,</span></span>
<span class="r-in"><span><span class="fl">1.188</span>, <span class="fl">2.074</span>, <span class="fl">21.262</span>, <span class="fl">4.206</span>,  <span class="fl">5.924</span>, <span class="fl">0.021</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">4</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># define marker names as rownames</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">intensity_matrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DAPI"</span>, <span class="st">"CD3"</span>, <span class="st">"CD4"</span>, <span class="st">"AMACR"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># define cell IDs as colnames</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">intensity_matrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cell_1"</span>, <span class="st">"Cell_2"</span>, <span class="st">"Cell_3"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Construct a dummy metadata (phenotypes, x/y coordinates)</span></span></span>
<span class="r-in"><span><span class="co"># the order of the elements in these vectors correspond to the cell order</span></span></span>
<span class="r-in"><span><span class="co"># in `intensity matrix`</span></span></span>
<span class="r-in"><span><span class="va">phenotypes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"OTHER"</span>,  <span class="st">"AMACR"</span>, <span class="st">"CD3,CD4"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">coord_x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">82</span>, <span class="fl">171</span>, <span class="fl">184</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">coord_y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">30</span>, <span class="fl">22</span>, <span class="fl">38</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">formatted_image</span> <span class="op">&lt;-</span> <span class="fu">format_image_to_spe</span><span class="op">(</span>intensity_matrix<span class="op">=</span><span class="va">intensity_matrix</span>,</span></span>
<span class="r-in"><span>phenotypes <span class="op">=</span> <span class="va">phenotypes</span>, coord_x <span class="op">=</span> <span class="va">coord_x</span>,coord_y <span class="op">=</span> <span class="va">coord_y</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Note: 4 rows removed due to NA intensity values.</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Anna Trigos, Yuzhou Feng, Tianpei Yang, Mabel Li, John Zhu, Volkan Ozcoban, Maria Doyle.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

