<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Formats an INFORM, HALO, Visum, CODEX, cellprofiler outputs into
  a singlecellexperiment class. Users can also input a generic format. The
  count assay stores the intensity level of every marker (rows) for every
  cell (columns). Cell phenotype, x and y coordinates and other properties
  are stored under colData. If INFORM format, the cell properties are
  Cell.Area, Nucleus.Area, Nucleus.Compactness, Nucleus.Axis.Ratio, and
  Cell.Axis.Ratio. If HALO format, the cell properties are Cell.Area,
  Nucleus.Area and Cytoplasm.Area."><title>format_image_to_sce — format_image_to_sce • SPIAT</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="format_image_to_sce — format_image_to_sce"><meta property="og:description" content="Formats an INFORM, HALO, Visum, CODEX, cellprofiler outputs into
  a singlecellexperiment class. Users can also input a generic format. The
  count assay stores the intensity level of every marker (rows) for every
  cell (columns). Cell phenotype, x and y coordinates and other properties
  are stored under colData. If INFORM format, the cell properties are
  Cell.Area, Nucleus.Area, Nucleus.Compactness, Nucleus.Axis.Ratio, and
  Cell.Axis.Ratio. If HALO format, the cell properties are Cell.Area,
  Nucleus.Area and Cytoplasm.Area."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SPIAT</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/introduction.html">Overview of the SPIAT package</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cancer-evolution/SPIAT/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>format_image_to_sce</h1>
      <small class="dont-index">Source: <a href="https://github.com/cancer-evolution/SPIAT/blob/HEAD/R/format_image_to_sce.R" class="external-link"><code>R/format_image_to_sce.R</code></a></small>
      <div class="d-none name"><code>format_image_to_sce.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Formats an INFORM, HALO, Visum, CODEX, cellprofiler outputs into
  a singlecellexperiment class. Users can also input a generic format. The
  count assay stores the intensity level of every marker (rows) for every
  cell (columns). Cell phenotype, x and y coordinates and other properties
  are stored under colData. If INFORM format, the cell properties are
  Cell.Area, Nucleus.Area, Nucleus.Compactness, Nucleus.Axis.Ratio, and
  Cell.Axis.Ratio. If HALO format, the cell properties are Cell.Area,
  Nucleus.Area and Cytoplasm.Area.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">format_image_to_sce</span><span class="op">(</span>
  format <span class="op">=</span> <span class="st">"INFORM"</span>,
  path <span class="op">=</span> <span class="cn">NULL</span>,
  markers <span class="op">=</span> <span class="cn">NULL</span>,
  locations <span class="op">=</span> <span class="cn">NULL</span>,
  dye_columns_interest <span class="op">=</span> <span class="cn">NULL</span>,
  intensity_columns_interest <span class="op">=</span> <span class="cn">NULL</span>,
  path_to_visium_coordinates <span class="op">=</span> <span class="cn">NULL</span>,
  path_to_codex_cell_phenotypes <span class="op">=</span> <span class="cn">NULL</span>,
  intensity_matrix <span class="op">=</span> <span class="cn">NULL</span>,
  phenotypes <span class="op">=</span> <span class="cn">NULL</span>,
  coord_x <span class="op">=</span> <span class="cn">NULL</span>,
  coord_y <span class="op">=</span> <span class="cn">NULL</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>format</dt>
<dd><p>String defining the software used for cell segmentation.
Options: "INFORM", "HALO", "Visium", "CODEX" "cellprolifer, or "general"</p></dd>
<dt>path</dt>
<dd><p>String of the path location of either HALO csv file, INFORM text
file or CODEX csv file. If 10X Visium transcriptomics data, path to the
"filtered_feature_bc_matrix" folder generated by CellRanger.</p></dd>
<dt>markers</dt>
<dd><p>For INFORM and HALO formats. Vector containing the markers
used for staining. Must match the order of the 'markers' parameter</p></dd>
<dt>locations</dt>
<dd><p>Vector containing the locations of markers used for
staining. Location can be either "Nucleus", "Cytoplasm" or "Membrane". This
is used to select the Intensity column and can be used instead of
intensity_columns_interest.</p></dd>
<dt>dye_columns_interest</dt>
<dd><p>For HALO formats. Use if locations is not
specified. Vector of names of the columns with the marker status (i.e.
those indicating 1 or 0 for whether the cell is positive or negative for
the marker). Column names must match the order of the 'markers' parameter.</p></dd>
<dt>intensity_columns_interest</dt>
<dd><p>For HALO formats. Use if locations is not
specified. Vector with the names of the columns with the level of each
marker. Column names must match the order of the 'markers' parameter.</p></dd>
<dt>path_to_visium_coordinates</dt>
<dd><p>For 10X Visium data. String of the path to
the "tissue_positions_list.csv" generated by CellRanger.</p></dd>
<dt>path_to_codex_cell_phenotypes</dt>
<dd><p>For CODEX data. String of the path to
the Cluster ID/Cell type file.</p></dd>
<dt>intensity_matrix</dt>
<dd><p>For the "general" format. A matrix of marker
intensities or gene expression where the column names are the Cell IDs, and
the rownames the marker or genes.</p></dd>
<dt>phenotypes</dt>
<dd><p>A vector of cell phenotypes in the same order in which they
appear in intensity_matrix. If no phenotypes available, then a vector of NA
can be used as input. Note that the combination of markers (e.g. CD3,CD4)
needs to be used instead of the cell type name (e.g. helper T cells)</p></dd>
<dt>coord_x</dt>
<dd><p>A vector with the X coordinates of the cells. The cells must
be in the same order as in the intensity_matrix.</p></dd>
<dt>coord_y</dt>
<dd><p>A vector with the Y coordinates of the cells. The cells must
be in the same order as in the intensity_matrix.</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A SingleCellExperiment object is returned</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Anna Trigos, Yuzhou Feng, Tom Yang, Mabel Li, John Zhu, Volkan Ozcoban, Maria Doyle.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

    </footer></div>

  

  

  </body></html>

