#' format_image_to_sce
#'
#' @description Formats an INFORM, HALO, Visum, CODEX, cellprofiler outputs into a singlecellexperiment class. Users can also input a generic format.
#' The count assay stores the intensity level of every marker (rows) for
#' every cell (columns). Cell phenotype, x and y coordinates and other properties are stored under colData. If INFORM format, the cell properties 
#' are Cell.Area, Nucleus.Area, Nucleus.Compactness, Nucleus.Axis.Ratio, and Cell.Axis.Ratio. If HALO format, the cell properties are Cell.Area, 
#' Nucleus.Area and Cytoplasm.Area.
#' 
#' @export
#' @param format String defining the software used for cell segmentation. Options: "INFORM", "HALO", "Visium", "CODEX" "cellprolifer, or "general"
#' @param path String of the path location of either HALO csv file, INFORM text file or CODEX csv file. If 10X Visium transcriptomics data,
#' path to the "filtered_feature_bc_matrix" folder generated by CellRanger
#' @param path_to_visium_coordinates For 10X Visium data. String of the path to the "tissue_positions_list.csv" generated by CellRanger
#' @param path_to_codex_cell_phenotypes For CODEX data. String of the path to the Cluster ID/Cell type file.
#' @param markers For INFORM and HALO formats. Vector containing the markers used for staining. Must match the order of the 'markers' parameter
#' @param locations Vector containing the locations of markers used for staining. Location can be either "Nucleus", "Cytoplasm" or "Membrane". 
#' This is used to select the Intensity column and can be used instead of intensity_columns_interest. 
#' @param dye_columns_interest For HALO formats. Use if locations is not specified. Vector of names of the columns with the marker status 
#' (i.e. those indicating 1 or 0 for whether the cell is positive or negative for the marker). 
#' Column names must match the order of the 'markers' parameter.
#' @param intensity_columns_interest For HALO formats. Use if locations is not specified. Vector with the names of the columns with the level of each marker.
#' Column names must match the order of the 'markers' parameter
#' @param intensity_matrix For the "general" format. A matrix of marker intensities or gene expression where the column names are the Cell IDs, and the rownames the marker or genes.
#' @param phenotypes A vector of cell phenotypes in the same order in which they appear in intensity_matrix. If no phenotypes available, then a vector of NA can be used as input. Note
#' that the combination of markers (e.g. CD3,CD4) needs to be used instead of the cell type name (e.g. helper T cells)
#' @param coord_x A vector with the X coordinates of the cells. The cells must be in the same order as in the intensity_matrix.
#' @param coord_y A vector with the Y coordinates of the cells. The cells must be in the same order as in the intensity_matrix.
#' @importFrom SingleCellExperiment SingleCellExperiment	
#' @importFrom SummarizedExperiment colData
#' @importFrom vroom vroom
#' @importFrom utils read.delim read.csv
#' @importFrom Seurat Read10X
#' @return A SingleCellExperiment object is returned

format_image_to_sce <- function(format = "INFORM", 
                                path = NULL, 
                                markers = NULL, 
                                locations = NULL, 
                                dye_columns_interest = NULL, 
                                intensity_columns_interest = NULL,
                                path_to_visium_coordinates = NULL,
                                path_to_codex_cell_phenotypes = NULL,
                                intensity_matrix = NULL,
                                phenotypes = NULL,
                                coord_x = NULL,
                                coord_y = NULL){
  
  # function to remove rows where intensity is NA
  remove_intensity_na <- function(intensity_columns) {
      n_before <- nrow(intensity_columns)
      intensity_columns <- na.omit(intensity_columns)
      n_after <- length(attributes(na.omit(intensity_columns))$row.names)
      n <- n_before - n_after
      message(sprintf("Note: %i rows removed due to NA intensity values.",n))
      return(intensity_columns)
  }
  
  #process the data based on data format
  if (format == "HALO"){
    #following is from format_HALO_new
    
    #read in the image using vroom for super fast import
    image <- vroom(path)
    
    #remove spaces from column names
    colnames(image) <- make.names(colnames(image))
    
    # if locations is specified, use location plus marker name to get the intensity and dye columns
    if (!is.null(locations)) {
      
      intensity_columns_interest <- character(length(markers))
      dye_columns_interest <- character(length(markers))
      i <- 1
      
      for (loc in locations) {
        if (loc == "Nucleus") {
          intensity_columns_interest[i] <- paste0("Dye.", i, ".Nucleus.Intensity")
          dye_columns_interest[i] <- paste0("Dye.", i, ".Positive.Nucleus")
        } else if (loc == "Cytoplasm") {
          intensity_columns_interest[i] <- paste0("Dye.", i, ".Cytoplasm.Intensity")
          dye_columns_interest[i] <- paste0("Dye.", i, ".Positive.Cytoplasm")
        } else if (loc == "Membrane") {
          intensity_columns_interest[i] <- paste0("Dye.", i, ".Membrane.Intensity")
          dye_columns_interest[i] <- paste0("Dye.", i, ".Positive.Membrane")
        } else {
          stop('Location incorrectly specified. Must be either "Nucleus", "Cytoplasm" or "Membrane"')
        }
        
        i <- i + 1
      }
      
      # if locations not used, get the intensity and dye columns from their specified names
    } else {
      
      #replace the spaces and non-alphanumeric characters as a '.' for column selection
      intensity_columns_interest <- gsub("[^[:alnum:]]", ".", intensity_columns_interest)
      dye_columns_interest <- gsub("[^[:alnum:]]", ".", dye_columns_interest)
    
      #CHECK - if image contains all the columns specified and vectors of same length
      image_colnames <- colnames(image)
      if (!all(intensity_columns_interest %in% image_colnames)) {
        stop("One or more Intensity_columns_interest not found in image")
      }
      if (!all(dye_columns_interest %in% image_colnames)) {
        stop("One or more dye_columns_interest not found in image")
      }
      marker_count <- length(markers)
      intensity_col_count <- length(intensity_columns_interest)
      dye_col_count <- length(dye_columns_interest)
      if (marker_count != intensity_col_count || marker_count != 
          dye_col_count || intensity_col_count != dye_col_count) {
        stop("The number of dyes, columns and markers do not match")
      }
    }
    
    #First remove all non-DAPI cells
    idx <- which(markers == "DAPI")
    
    #CHECK - if DAPI is in the dyes
    if (length(idx) == 0) {
      stop("Please include DAPI in the markers")
    }
    
    DAPI_col_name <- dye_columns_interest[idx]
    DAPI_non_zero_rows <- which(image[,DAPI_col_name] != 0)
    image <- image[DAPI_non_zero_rows,]
    
    #extract intensities
    intensity_of_markers <- image[, c("Object.Id", intensity_columns_interest)]
    colnames(intensity_of_markers) <- c("Object.Id", markers)
    intensity_of_markers[intensity_of_markers == "#N/A"] <- NA
    intensity_of_markers <- remove_intensity_na(intensity_of_markers)
    intensity_of_markers <- apply(intensity_of_markers, 2, function(x){
      as.numeric(as.character(x))
    })
    
    # remove intensity NA rows from image
    image <- subset(image, Object.Id %in% intensity_of_markers[, "Object.Id"])
    intensity_of_markers <- intensity_of_markers[ , !(colnames(intensity_of_markers) == "Object.Id")]
  
    
    #get the intensity status columns
    intensity_status_cols <- image[,dye_columns_interest]
    colnames(intensity_status_cols) <- markers
    
    #grab relevant columns
    cell_properties_cols <- colnames(image)[grep("Nucleus.Area|Cytoplasm.Area|Cell.Area", colnames(image))]
    image <- image[,c("Object.Id", "XMin", "XMax", "YMin", "YMax", cell_properties_cols)]
    
    #rename Object.ID to Cell.ID
    colnames(image)[1] <- "Cell.ID"
    
    #add "Cell_" in front of Cell.ID
    image$Cell.ID <- paste("Cell_", image$Cell.ID, sep="")
    
    #add averaged X and Y position
    image$Cell.X.Position <- (image$XMin + image$XMax)/2
    image$Cell.Y.Position <- (image$YMin + image$YMax)/2
    
    #start reading in the Phenotypes of every cell
    intensity_status_cols$Phenotype <- ""
    for (marker in markers) {
      if (marker == "DAPI") {
        phenotype <- "OTHER,"
      } else {
        phenotype <- paste(marker, ",", sep = "")
      }
      
      #get the row idx of the cells that express the specific marker, and paste the phenotype
      rows_true_exp <- which(intensity_status_cols[,marker] != 0)
      if (length(rows_true_exp) != 0) {
        intensity_status_cols[rows_true_exp,]$Phenotype <- paste(intensity_status_cols[rows_true_exp,]$Phenotype, phenotype, sep="")
      }
    }
    
    #now clean the phenotype column
    if (nrow(intensity_status_cols[intensity_status_cols$Phenotype == "OTHER,", ]) != 0) {
      intensity_status_cols[intensity_status_cols$Phenotype == "OTHER,", ]$Phenotype <- "OTHER"
    }
    intensity_status_cols$Phenotype <- gsub("OTHER,", "", intensity_status_cols$Phenotype)
    intensity_status_cols$Phenotype <- gsub(",OTHER", "", intensity_status_cols$Phenotype)
    intensity_status_cols$Phenotype <- gsub(",$", "", intensity_status_cols$Phenotype)
    
    #grab the phenotype column and cbind to image
    phenotype_column <- data.frame(intensity_status_cols$Phenotype)
    colnames(phenotype_column) <- "Phenotype"
    
    image <- cbind(image, phenotype_column)
    
    image$Phenotype <- as.character(image$Phenotype)
    
    image <- image[,c("Cell.ID", "Phenotype", "Cell.X.Position", "Cell.Y.Position", cell_properties_cols)]
    
    #create the formatted_data with intensity levels
    formatted_data <- cbind(image, intensity_of_markers)
    
    #now create the SCE object...
    #grab the intensity level, markers and cell IDs
    assay_data <- formatted_data[,c(markers)]
    
    #transpose the matrix so every column is a cell and every row is a marker
    assay_data_matrix <- as.matrix(assay_data)
    colnames(assay_data_matrix) <- NULL
    rownames(assay_data_matrix) <- NULL
    assay_data_matrix_t <- t(assay_data_matrix)
    
    sce <- SingleCellExperiment(assays = list(counts = assay_data_matrix_t))
    
    rownames(sce) <- markers
    colnames(sce) <- formatted_data[,"Cell.ID"]
    
    #Assign the phenotype, X and Y positions and cell property columns as the colData
    metadata_columns <- formatted_data[ ,c("Phenotype", "Cell.X.Position", "Cell.Y.Position", cell_properties_cols)]
    SummarizedExperiment::colData(sce) <- cbind(SummarizedExperiment::colData(sce), metadata_columns)
    
  } else if (format == "INFORM"){
    
    ###following codes are from format_INFORM
    
    #read in the image file
    image <- vroom(path)
    #remove all rows with empty phenotype/no markers
    image <- image[image$Phenotype != "",]
    image <- image[!is.na(image$Phenotype), ]
    
    if (!is.null(locations)) {
    
      # add the location of interest to each marker
      names_to_match <- paste(locations, markers, sep=" ")
      
      # get all the mean intensity column names in the file
      intensity_col_all <- colnames(image)[grepl("Mean \\(Normalized Counts, Total Weighting\\)", colnames(image))]
      
      # get the intensity column name for each marker
      intensity_columns_interest <- character(length(markers))
      i <- 1
      
      for (name in names_to_match) {
        intensity_columns_interest[i] <- intensity_col_all[grepl(name, intensity_col_all)]
        i <- i + 1
      }
    
      } else {
    
        #CHECK - if image contains all the columns specified and vectors of same length
        image_colnames <- colnames(image)
        if (!all(intensity_columns_interest %in% image_colnames)) {
          stop("One or more Intensity_columns_interest not found in image")
        }
        marker_count <- length(markers)
        intensity_col_count <- length(intensity_columns_interest)
        if (marker_count != intensity_col_count) {
          stop("The number of dyes and columns does not match")
        }
    }
    
    ###added: extract intensities
    intensity_of_markers <- image[,c("Cell ID", intensity_columns_interest)]
    colnames(intensity_of_markers) <- c("Cell ID", markers)
    intensity_of_markers[intensity_of_markers == "#N/A"] <- NA
    intensity_of_markers <- remove_intensity_na(intensity_of_markers)
    intensity_of_markers <- apply(intensity_of_markers, 2, function(x){
      as.numeric(as.character(x))
    })
    
    # remove intensity NA rows from image
    image <- subset(image, `Cell ID` %in% intensity_of_markers[, "Cell ID"])
    intensity_of_markers <- intensity_of_markers[ , !(colnames(intensity_of_markers) == "Cell ID")]
    
    #extract the columns of interest and discard the rest
    colnames(image) <- make.names(colnames(image))
    
    # shorten column names of some properties 
    names(image)[names(image)=="Entire.Cell.Area..pixels."] <- "Cell.Area"
    names(image)[names(image)=="Nucleus.Area..pixels."] <- "Nucleus.Area"
    names(image)[names(image)=="Entire.Cell.Axis.Ratio"] <- "Cell.Axis.Ratio"

    cell_properties_cols <- c("Cell.Area","Nucleus.Area",
                              "Nucleus.Compactness","Nucleus.Axis.Ratio",
                              "Cell.Axis.Ratio")
    image <- image[ ,c("Cell.ID", "Phenotype", "Cell.X.Position",
                       "Cell.Y.Position", cell_properties_cols)]
    
    #add 'cell_' to the start of the objectId
    image$Cell.ID <- paste0("Cell_", image$Cell.ID)
    
    #reformat the phenotype into "marker1, marker2..."
    for (marker in markers) {
      marker_positive <- paste0(marker, "+", sep="")
      marker_negative <- paste0(marker, "-", sep="")
      
      marker_replacement <- paste0(marker,",", sep="")
      
      image$Phenotype <- gsub(marker_positive, marker_replacement, image$Phenotype, fixed=TRUE)
      image$Phenotype <- gsub(marker_negative, "", image$Phenotype, fixed=TRUE)
      
    }
    
    #remove the comma at the end
    image$Phenotype <- gsub(",$","", image$Phenotype)
    
    #create the formatted_data with intensity levels
    formatted_data <- cbind(image, intensity_of_markers)
    
    #now create the SCE object...
    #grab the intensity level, markers and cell IDs
    assay_data <- formatted_data[,c(markers)]
    
    #transpose the matrix so every column is a cell and every row is a marker
    assay_data_matrix <- as.matrix(assay_data)
    colnames(assay_data_matrix) <- NULL
    rownames(assay_data_matrix) <- NULL
    assay_data_matrix_t <- t(assay_data_matrix)
    
    sce <- SingleCellExperiment(assays = list(counts = assay_data_matrix_t))
    
    rownames(sce) <- markers
    colnames(sce) <- formatted_data[,"Cell.ID"]
    
    #Assign the phenotype, X and Y positions and cell property columns as the colData
    metadata_columns <- formatted_data[ ,c("Phenotype", "Cell.X.Position", "Cell.Y.Position", cell_properties_cols)]
    SummarizedExperiment::colData(sce) <- cbind(SummarizedExperiment::colData(sce), metadata_columns)
    
  } else if (format == "Visium"){
    
    raw_image <- Read10X(data.dir = path)
    expression <- as.matrix(raw_image)
    cells <- colnames(expression)
    
    location <- read.delim(path_to_visium_coordinates, header=FALSE, sep=",")
    location$V1 <- gsub("-1", "", location$V1)
    
    #V2 are the empty spots. Removing from the datasets
    location <- location[location$V2 == 1,] 
    location$V2 <- NULL
    location$V5 <- NULL
    location$V6 <- NULL
    
    colnames(location) <- c("Cell", "Cell.Y.Position", "Cell.X.Position")
    location$Phenotype <- NA
    
    expression <- expression[,colnames(expression) %in% location$Cell]
    
    sce <- SingleCellExperiment(assays = list(counts = expression))
    
    location <- location[match(colnames(expression), location$Cell),]
    
    metadata_columns <- location[,c("Phenotype", "Cell.X.Position", "Cell.Y.Position")]
    SummarizedExperiment::colData(sce) <- cbind(SummarizedExperiment::colData(sce), metadata_columns)
    
  } else if(format == "CODEX") { 
    
    phenotypes <- read.delim(path_to_codex_cell_phenotypes,header=FALSE)
    colnames(phenotypes) <- c("Cluster_ID", "Cell_type")
    
    data <- read.delim(path, sep=",")
    data$Cell_type <- phenotypes$Cell_type[match(data$Imaging.phenotype.cluster.ID,
                                                 phenotypes$Cluster_ID)]
    data$Imaging.phenotype.cluster.ID <- NULL
    data$niche.cluster.ID <- NULL  
    data$sample_Xtile_Ytile <- NULL
    data$Z.Z <- NULL
    
    coordinates <- data[,c("X.X", "Y.Y")]
    data$X.X <- NULL
    data$Y.Y <- NULL
    phenotype <- data$Cell_type
    data$Cell_type <- NULL
    rownames(data) <- paste("Cell", rownames(data), sep="_")
    data <- t(data)
    
    sce <- SingleCellExperiment(assays = list(counts = data))
    
    metadata_columns <- data.frame(Phenotype = phenotype,
                                   Cell.X.Position = coordinates$X.X,
                                   Cell.Y.Position = coordinates$Y.Y)
    SummarizedExperiment::colData(sce) <- cbind(SummarizedExperiment::colData(sce), metadata_columns)
    
  } else if(format == "cellprofiler"){
      # Due to the implementation, rename DNA to DAPI before usage. 
      
      if (substr(path, nchar(path)-3+1, nchar(path)) != "csv"){
        stop("Please convert your file from a cpout to a csv format")
      }
      
      #read in the image
      image <- read.csv(path)
      
      #CHECK - if image contains all the columns specified and vectors of same length
      image_colnames <- colnames(image)
      if (!all(intensity_columns_interest %in% image_colnames)) {
        stop("One or more Intensity_columns_interest not found in image")
      }
      
      marker_count <- length(markers)
      intensity_col_count <- length(intensity_columns_interest)
      
      if (marker_count != intensity_col_count) {
        stop("The number of columns and markers do not match")
      }
      
      #extract intensities
      intensity_of_markers <- image[,c("ObjectNumber", intensity_columns_interest)]
      colnames(intensity_of_markers) <- c("ObjectNumber", markers)
      intensity_of_markers[intensity_of_markers == "#N/A"] <- NA
      intensity_of_markers <- remove_intensity_na(intensity_of_markers)
      intensity_of_markers <- apply(intensity_of_markers, 2, function(x){
        as.numeric(as.character(x))
      })
    
      # remove intensity NA rows from image
      image <- subset(image, ObjectNumber %in% intensity_of_markers[, "ObjectNumber"])
      intensity_of_markers <- intensity_of_markers[ , !(colnames(intensity_of_markers) == "ObjectNumber")]
      
      #grab relevant columns
      image <- image[,c("ObjectNumber", "Location_Center_X", "Location_Center_Y")]
      
      #rename Object.ID to Cell.ID
      colnames(image)[colnames(image) %in% c("ObjectNumber")] <- c("Cell.ID")
      
     
      
      #add "Cell_" in front of Cell.ID
      image$Cell.ID <- paste("Cell_", image$Cell.ID, sep="")
      
      colnames(image)[colnames(image) %in% c("Location_Center_X", "Location_Center_Y")] <- c("Cell.X.Position", "Cell.Y.Position")
      
      image$Phenotype <- NA
      image <- image[,c("Cell.ID", "Phenotype", "Cell.X.Position", "Cell.Y.Position")]
      
      #create the formatted_data with intensity levels
      formatted_data <- cbind(image, intensity_of_markers)
      
      #now create the SCE object...
      #grab the intensity level, markers and cell IDs
      assay_data <- formatted_data[,c(markers)]
      
      #transpose the matrix so every column is a cell and every row is a marker
      assay_data_matrix <- as.matrix(assay_data)
      colnames(assay_data_matrix) <- NULL
      rownames(assay_data_matrix) <- NULL
      assay_data_matrix_t <- t(assay_data_matrix)
      
      sce <- SingleCellExperiment(assays = list(counts = assay_data_matrix_t))
      
      rownames(sce) <- markers
      colnames(sce) <- formatted_data[,"Cell.ID"]
      
      #Assign the phenotype, X and Y positions and cell property columns as the colData
      metadata_columns <- formatted_data[ ,c("Phenotype", "Cell.X.Position", "Cell.Y.Position")]
      SummarizedExperiment::colData(sce) <- cbind(SummarizedExperiment::colData(sce), metadata_columns)
      
      
      
      
  } else if(format == "general"){
    intensity_matrix <- remove_intensity_na(intensity_matrix)
    sce <- SingleCellExperiment(assays = list(counts = intensity_matrix))
    
    rownames(sce) <- rownames(intensity_matrix)
    colnames(sce) <- colnames(intensity_matrix)
    
    metadata_columns <- data.frame(Phenotype = phenotypes,
                                   Cell.X.Position = coord_x,
                                   Cell.Y.Position = coord_y)
    SummarizedExperiment::colData(sce) <- cbind(SummarizedExperiment::colData(sce), metadata_columns)
    
  } else{
    stop("Please enter a valid format")
  }
  
  return(sce)
}
